### DISCUSSION TITLE
The mystery of 'undefined'

### DISCUSSION THREAD

**Discussion by ttsoares** [vercel/next.js#49225] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #44-Ubuntu SMP Wed Jun 22 14:20:53 UTC 2022
    Binaries:
      Node: 18.16.0
      npm: 9.6.4
      Yarn: 1.22.19
      pnpm: 7.17.1
    Relevant packages:
      next: 13.3.5-canary.7
      eslint-config-next: 13.3.2
      react: 18.2.0
      react-dom: 18.2.0
```

### Link to the code that reproduces this issue

https://github.com/ttsoares/whispers

### To Reproduce

Locally or in Vercel the problem is the same.

1) Clone the repository
2) npm run build

### Describe the Bug

The error:
**TypeError: Cannot destructure property 'name' of 'ownerUser' as it is undefined.**

The code:
`
const ownerUser = usersArray.find((elm) => Number(elm.id) === ownerId);

console.log(ownerUser);

const { name } = ownerUser;
`

The log print the content of the object ownerUser:
_id: 1
name: "qqq"
password: "202cb962ac59075b964b07152d234b70"
type: "wss"_

Also this do no build:
`const ownerUser = usersArray.find((elm) => Number(elm.id) === ownerId).name`

**The system runs normally with 'npm run dev'**

### Expected Behavior

Expect to build and all is running fine with run dev.

### Which browser are you using? (if relevant)

Firefox

### How are you deploying your application? (if relevant)

Vercel

**Comment by ttsoares** [comment]
**I'm a NextJS newbie**
After hours of reading and trying... it seems that React code is no really the same as a NextJS code !

To solve this particular case, I did this:
` 
 const router = useRouter();
  const [ownerId, setOwnerId] = useState(0);
  const [ownerName, setOwnerName] = useState("");

  useEffect(() => {
    if (router.isReady) {
      setOwnerId(Number(router.query.dash[0]));
      setOwnerName(router.query.dash[1]);
    }
  }, [router.isReady, router.query.dash]);
`

**Comment by icyJoseph** [comment]
Hi,

Wasn't the issue originally that, on first render, the query object is empty? And then immediately Next.js triggers a re-render with the query filled in?

> Pages that are statically optimized by [Automatic Static Optimization](https://nextjs.org/docs/pages/building-your-application/rendering/automatic-static-optimization) will be hydrated without their route parameters provided, i.e query will be an empty object ({}).

https://nextjs.org/docs/pages/building-your-application/rendering/automatic-static-optimization#how-it-works

**Comment by ttsoares** [comment]
Yes. I'm learning NextJS by doing... and this behavior fooled me completely
!
Already closed the "bug" report.
Sorry. My bad.

Em sex., 5 de mai. de 2023 às 09:41, Joseph ***@***.***>
escreveu:
